internal/daemon/cache/search.go

```diff
@@ -0,0 +1,155 @@
+package cache
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/boundary/api/sessions"
+	"github.com/hashicorp/boundary/api/targets"
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/util"
+	"github.com/hashicorp/go-bexpr"
+)
+
+// SearchParams contains the paramaters for searching in the cache
+type SearchParams struct {
+	// the name of the resource. eg. "targets" or "sessions"
+	Resource string
+	// the auth token id for the user id that has resources synced to the cache
+	AuthTokenId string
+	// the optional mql query to use when searching the resources.
+	Query string
+	// the optional bexpr filter string that all results will be filtered by
+	Filter string
+}
+
+// SearchResult returns the results from searching the cache.
+type SearchResult struct {
+	Targets  []*targets.Target
+	Sessions []*sessions.Session
+}
+
+// SearchService is a domain service that can search across all resources in the
+// cache.
+type SearchService struct {
+	repo                *Repository
+	searchableResources map[string]searcher
+}
+
+func NewSearchService(ctx context.Context, repo *Repository) (*SearchService, error) {
```

Comment:
could repo be an interface with a defined set of receiver funcs?  I'm not sure
it needs to depend on the entire concrete type.   I didn't look to see what's
not used from the concrete type.